synapse/handlers/directory.py

```diff
@@ -96,6 +94,21 @@ async def _create_association(
             room_alias, room_id, servers, creator=creator
         )
 
+    async def _get_servers_known_to_be_in_room(self, room_id: str) -> Set[str]:
+        if await self.store.is_partial_state_room(room_id):
+            # We don't have the full list of hosts in the room yet, but we do have
+            # some of them from the `/send_join` response. Use those to best-effort
+            # approximate the full list of servers in the room.
+            servers = set(await self.store.get_partial_state_servers_at_join(room_id))
+
+            # We are also in the room.
+            servers.add(self.server_name)
+        else:
+            servers = await self._storage_controllers.state.get_current_hosts_in_room(
+                room_id
+            )
+        return servers
+
```

Comment:
We have an existing method, `self._storage_controllers.state.get_current_hosts_in_room_or_partial_state_approximation` that is more or less equivalent. Is there a reason we're avoiding it (apart from the awful name, which is my fault)?

Reply:
Only that I didn't know it existed. Will update, thanks.